### PR TITLE
feat: Add range opcode optimization

### LINKED ISSUE #218: Range gate optimization
### Problem

After this PR in barretenberg is merged (https://github.com/AztecProtocol/barretenberg/pull/369) an inefficiency has been introduced. Whenever there is an overlapping range check on the same witness we have to add an arithmetic gate.  

### Proposed solution

We can avoid this additional arithmetic gate by optimizing out redundant range constraints. We can remove the less restrictive range checks and keep the most restrictive. For example, if we range check the value at witness index 90 with a bit size of 4 and 8 in two different constraints we can remove the range constraint with bit size of 8. 

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by kevaundray**
# Related issue(s)

Resolves #218 

(If it does not already exist, first create a GitHub issue that describes the problem this Pull Request (PR) solves before creating the PR and link it here.)

Resolves (link to issue)

# Description

## Summary of changes

(Describe the changes in this PR. Point out breaking changes if any.)

## Dependency additions / changes

(If applicable.)

## Test additions / changes

(If applicable.)

# Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.

# Additional context

(If applicable.)

### CODE REVIEW

**Review by TomAFrench — APPROVED** [review]
Generally looks good.

If we're being lazy about this rather than having the two passes be
1. Build a mapping from witness to strictest `num_bits`.
2. Iterate through opcodes, get minimal `num_bits`, check opcode hasn't already been added for witness already, retain opcode and mark witness as seen

We could change it to be:

1. Build a mapping from witness to range opcode index with strictest `num_bits`.
1.a. convert mapping to a HashSet of opcode indices to retain.
2. Iterate through opcode discarding any range opcodes with indices not in the set.

The benefits probably aren't worth the extra time to change however.

File: acvm/src/compiler/optimizers/mod.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
 mod general;
+mod redundant_range;
 
 pub use general::GeneralOpt as GeneralOptimizer;
+pub use redundant_range::RangeOptimizer;
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply acvm/src/compiler/optimizers/mod.rs:5]: Fair point, lets make it non-public

**Comment by kevaundray** [comment]
> Generally looks good.
> 
> If we're being lazy about this rather than having the two passes be
> 
> 1. Build a mapping from witness to strictest `num_bits`.
> 2. Iterate through opcodes, get minimal `num_bits`, check opcode hasn't already been added for witness already, retain opcode and mark witness as seen
> 
> We could change it to be:
> 
> 1. Build a mapping from witness to range opcode index with strictest `num_bits`.
>    1.a. convert mapping to a HashSet of opcode indices to retain.
> 2. Iterate through opcode discarding any range opcodes with indices not in the set.
> 
> The benefits probably aren't worth the extra time to change however.

Yeah I was thinking of being smart by not doing a double pass, but I just went with the simplest version now and then if we see it as a bottle-neck, can change to a more optimized version

**Comment by TomAFrench** [comment]
> Yeah I was thinking of being smart by not doing a double pass

Actually yeah, we don't need to do a double pass at all. I was thinking you were going through the effort of not just bunging the range opcodes at the end as ideally we maintain the ordering of opcodes (as this minimises loops through the circuit looking for solvable opcodes) but range opcodes don't have any outputs so this isn't a concern.

**Pull request merged**